### PR TITLE
autogen: refactor

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

PR https://github.com/pmodels/mpich/pull/5645 contained some cleanup of `autogen.sh` that are orthogonal to the original goal. This PR is a result of separating the refactoring commits plus some additional commits to make the cleanup more complete.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/most

File: autogen.sh [hunk 1]
```diff
@@ -207,6 +207,56 @@ fn_getcvars() {
...
+        _patch_libtool $_dir/confdb/libtool.m4 big-sur.patch
+        if test "$do_bindings" = "yes" ; then
+            _patch_libtool $_dir/confdb/libtool.m4 darwin-ifort.patch
+            _patch_libtool $_dir/confdb/libtool.m4 oracle-fort.patch
+            _patch_libtool $_dir/confdb/libtool.m4 flang.patch
+            _patch_libtool $_dir/confdb/libtool.m4 arm-compiler.patch
+            _patch_libtool $_dir/confdb/libtool.m4 ibm-xlf.patch
+        fi
+
+        if test "$_patch_successful" = "yes" ; then
```
[the remark this answers is not in the document]
Comment (hzhou) [reply autogen.sh:579]: The flag is for whether to rerun `autoconf`, so as long as one patch is applied, it needs it to be "yes"

File: autogen.sh [hunk 2]
```diff
@@ -706,20 +706,13 @@ for arg in "$@" ; do
...
 	    opt=`echo A$arg | sed -e 's/^A--*do=//'`
-	    case $opt in 
-		build-configure|configure) opt=build_configure ;;
-	    esac
-	    var=do_$opt
-
-	    # Check that this opt is known
-	    eval oldval=\$"$var"
-	    if [ -z "$oldval" ] ; then
+            if test $(type fn_$opt) | grep -q 'function' ; then
```
[the remark this answers is not in the document]
Comment (hzhou) [reply autogen.sh:912]: The `type` the command is in POSIX (https://www.unix.com/man-page/posix/1p/type/), but many options are bash extensions. Here, I opted to use it without any options, hopefully is more portable.

[the remark this answers is not in the document]
Comment (raffenet) [reply autogen.sh:912]: However when I try this, I seem to get a syntax error:
```
raffenet@mcswl035:main/ (2111_autogen*) $ ./autogen.sh --do=f77
Verifying the location of autogen.sh... done
./autogen.sh: line 912: test: too many arguments
-do=f77 is unrecognized
```

**Review by raffenet — COMMENTED** [review]
Looks mostly good, but the `-do=<step>` functionality might have an issue.

**Comment by hzhou** [comment]
> but the -do=<step> functionality might have an issue.

Good catch! I should've tested it. Pushed the fix now.  The patch was:
```
-            if test $(type fn_$opt) | grep -q 'function' ; then
+            if type fn_$opt | grep -q 'function' ; then
```

**Comment by raffenet** [comment]
> > but the -do= functionality might have an issue.
> 
> Good catch! I should've tested it. Pushed the fix now. The patch was:
> 
> ```
> -            if test $(type fn_$opt) | grep -q 'function' ; then
> +            if type fn_$opt | grep -q 'function' ; then
> ```

Yup, that works for me.

**Review by raffenet — APPROVED** [review]

**Pull request merged**